cocotb/regression.py

```diff
@@ -229,63 +230,111 @@ def next_test(self):
 
+        # score test
+        result_pass, sim_failed = self.score_test(test, test._outcome)
+
+        # Save results
+        self._store_test_result(test.__module__, test.__name__, result_pass, sim_time_ns, real_time, ratio_time)
+        if not result_pass:
+            self.xunit.add_failure()
+            self.failures += 1
+
+        # Fail if required
+        if sim_failed:
+            self.tear_down()
+
+        self.execute()
+
+    def _init_test(self, test_func):
+        test_init_outcome = cocotb.outcomes.capture(test_func, self._dut)
+
+        if isinstance(test_init_outcome, cocotb.outcomes.Error):
+            self.log.error("Failed to initialize test %s" % test_func.name, exc_info=True)
+            self.xunit.add_testcase(name=test_func.name,
+                                    classname=test_func.__module__,
+                                    time="0.0",
+                                    sim_time_ns="0.0",
+                                    ratio_time="0.0")
+            result_pass, sim_failed = self.score_test(test_func, test_init_outcome.without_frames(['capture']))
+            # Save results
+            self._store_test_result(test_func.__module__, test_func.__name__, result_pass, 0.0, 0.0, 0.0)
+            if not result_pass:
+                self.xunit.add_failure()
+                self.failures += 1
+            # Fail if required
+            if sim_failed:
+                self.tear_down()
+
+        else:
+            test = test_init_outcome.get()
+            if test.skip:
+                self.log.info("Skipping test %s" % test_func.name)
+                self.xunit.add_testcase(name=test_func.name,
+                                        classname=test.module,
+                                        time="0.0",
+                                        sim_time_ns="0.0",
+                                        ratio_time="0.0")
+                self.xunit.add_skipped()
+                self.skipped += 1
+                self._store_test_result(test.module, test_func.name, None, 0.0, 0.0, 0.0)
+            else:
+                self._queue.append(test)
+                self.ntests += 1
+
+    def score_test(self, test, outcome):
+
         # Helper for logging result
         def _result_was():
             result_was = ("{} (result was {})".format
-                          (test.funcname, result.__class__.__name__))
+                          (test.__name__, result.__class__.__name__))
             return result_was
 
+        # scoring outcomes
         result_pass = True
+        sim_failed = False
 
-        # check what exception the test threw
         try:
-            test._outcome.get()
+            outcome.get()
         except Exception as e:
             if sys.version_info >= (3, 5):
-                result = remove_traceback_frames(e, ['handle_result', 'get'])
+                result = remove_traceback_frames(e, ['score_test', 'get'])
```

Comment:
Don't forget to change these with the function name
```suggestion
                result = remove_traceback_frames(e, ['_score_test', 'get'])
```